clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -314,68 +369,92 @@ public void poll(long timeout) throws IOException {
+        }
+        try {
+            for (SelectionKey key : inHandlingOrder) {
+                keysHandled.add(key);
+                KafkaChannel channel = channel(key);
+
+                // register all per-connection metrics at once
+                sensors.maybeRegisterConnectionMetrics(channel.id());
+                if (idleExpiryManager != null)
+                    idleExpiryManager.update(channel.id(), currentTimeNanos);
+
+                try {
+
+                    /* complete any connections that have finished their handshake (either normally or immediately) */
+                    if (isImmediatelyConnected || key.isConnectable()) {
+                        if (channel.finishConnect()) {
+                            this.connected.add(channel.id());
+                            this.sensors.connectionCreated.record();
+                            SocketChannel socketChannel = (SocketChannel) key.channel();
+                            log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
+                                    socketChannel.socket().getReceiveBufferSize(),
+                                    socketChannel.socket().getSendBufferSize(),
+                                    socketChannel.socket().getSoTimeout(),
+                                    channel.id());
+                        } else
+                            continue;
+                    }
 
-                /* if channel is ready read from any connections that have readable data */
-                if (channel.ready() && key.isReadable() && !hasStagedReceive(channel)) {
-                    NetworkReceive networkReceive;
-                    while ((networkReceive = channel.read()) != null)
-                        addToStagedReceives(channel, networkReceive);
-                }
+                    /* if channel is not ready finish prepare */
+                    if (channel.isConnected() && !channel.ready())
+                        channel.prepare();
+
+                    /* if channel is ready read from any connections that have readable data */
+                    if (channel.ready() && key.isReadable() && !hasStagedReceive(channel)) {
+                        NetworkReceive networkReceive;
+                        while ((networkReceive = channel.read()) != null)
+                            addToStagedReceives(channel, networkReceive);
+                        if (channel.hasBytesBuffered()) {
+                            //this channels has bytes enqueued in intermediary buffers that we could not read
+                            //(possibly because no memory). it may be the case that the underlying socket will
+                            //not come up in the next poll() and so we need to remember this channel for the
+                            //next poll call otherwise data may be stuff in said buffers forever.
+                            keysWithBytesBuffered.add(key);
+                        }
```

Comment:
@rajinisivaram - i've introduced a simple (relatively...) notion of progress made to try and prevent the tight loop you pointed out.